### ISSUE TITLE
`install_python()` fails on M1 Mac

### ISSUE THREAD

**Issue by jz-stat** [rstudio/reticulate#1350]
I was following the codes on https://tensorflow.rstudio.com/install/#installation.

```
install.packages("tensorflow")
library(reticulate)
path_to_python <- install_python()
```

It then gives me the errors:

```
> path_to_python <- install_python()
+ '/Users/jz/Library/Application Support/r-reticulate/pyenv/libexec/pyenv' install --skip-existing 3.9.16
/Users/jz/Library/Application Support/r-reticulate/pyenv/libexec/pyenv-latest: line 39: printf: write error: Broken pipe
Downloading openssl-1.1.1q.tar.gz...
-> https://www.openssl.org/source/openssl-1.1.1q.tar.gz
cat: /Users/jz/Library/Application: No such file or directory
cat: Support/r-reticulate/pyenv/plugins/python-build/bin/../share/python-build/patches/3.9.16/openssl-1.1.1q/openssl_1.1.1q_fix_c_include.patch: No such file or directory

BUILD FAILED (OS X 13.2.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/j3/bmhzpkjx5s9chbrjt7wf0rlm0000gn/T/python-build.20230309093409.68869
Results logged to /var/folders/j3/bmhzpkjx5s9chbrjt7wf0rlm0000gn/T/python-build.20230309093409.68869.log

Last 10 log lines:
/var/folders/j3/bmhzpkjx5s9chbrjt7wf0rlm0000gn/T/python-build.20230309093409.68869 ~
/var/folders/j3/bmhzpkjx5s9chbrjt7wf0rlm0000gn/T/python-build.20230309093409.68869/openssl-1.1.1q /var/folders/j3/bmhzpkjx5s9chbrjt7wf0rlm0000gn/T/python-build.20230309093409.68869 ~
Error: installation of Python 3.9.16 failed
```

Here is my ```sessionInfo()```.

```
> sessionInfo()
R version 4.2.2 (2022-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.2.1

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] reticulate_1.28

loaded via a namespace (and not attached):
 [1] compiler_4.2.2  Matrix_1.5-3    tools_4.2.2     withr_2.5.0     rappdirs_0.3.3 
 [6] Rcpp_1.0.10     grid_4.2.2      jsonlite_1.8.4  png_0.1-8       lattice_0.20-45
```

I tried to search for info on the Broken Pipeline error online, but I couldn't find any in RStudio IDE. I tried to see what's going on with the working tree at /var/... , but I couldn't even find them in my MacBook.

Any help is appreciated. Thank you!

**Comment by kevinushey** [comment]
```
cat: /Users/jz/Library/Application: No such file or directory
cat: Support/r-reticulate/pyenv/plugins/python-build/bin/../share/python-build/patches/3.9.16/openssl-1.1.1q/openssl_1.1.1q_fix_c_include.patch: No such file or directory
```

Looks like a "space in path" issue.

**Comment by t-kalinowski** [comment]
Thanks for reporting! It looks like there is a bug upstream in pyenv, where during compilation of openssl, a file path with a space in it is not quoted properly for the shell. We will have to investigate and figure out a fix.

In the interim, you could do the following:
- make sure `brew` is installed on your system. The presence of brew would enable pyenv to skip compilation of openssl.
- force a refresh of pyenv by calling `reticulate:::rm_all_reticulate_state()`
- Install Python manually by downloading the prebuilt `.pkg` from www.python.org, then use the path to that executable as your `path_to_python`.
- use `install_miniconda()` instead of `install_python()`.

Since your goal is to install TensorFlow on an M1 Mac, not that today, `install_miniconda()` is the only officially supported option for Tensorflow on M1 Macs.

```r
reticulate::install_miniconda()
tensorflow::install_tensorflow()
```

Some links relevant to TensorFlow on M1 Macs

- https://tensorflow.rstudio.com/install/local_gpu (macOS tab)
- https://tensorflow.rstudio.com/reference/tensorflow/install_tensorflow#apple-silicon

**Comment by zjl19902003** [comment]
> Thanks for reporting! It looks like there is a bug upstream in pyenv, where during compilation of openssl, a file path with a space in it is not quoted properly for the shell. We will have to investigate and figure out a fix.
> 
> In the interim, you could do the following:
> 
> * make sure `brew` is installed on your system (https://brew.sh/). The presence of brew would enable pyenv to skip compilation of openssl.
> * force a refresh of pyenv by calling `reticulate:::rm_all_reticulate_state()` and try `install_python()` again.
> * Install Python manually by downloading the prebuilt `.pkg` from [www.python.org](http://www.python.org), then use the path to that executable as your `path_to_python`.
> * use `install_miniconda()` instead of `install_python()`.
> 
> Since your goal is to install TensorFlow on an M1 Mac, not that today, `install_miniconda()` is the only officially supported option for Tensorflow on M1 Macs.
> 
> ```r
> reticulate::install_miniconda()
> tensorflow::install_tensorflow()
> ```
> 
> Some links relevant to TensorFlow on M1 Macs
> 
> * https://tensorflow.rstudio.com/install/local_gpu (macOS tab)
> * https://tensorflow.rstudio.com/reference/tensorflow/install_tensorflow#apple-silicon

Thanks! I was able to follow your guidance and seemed to successfully install TensorFlow. However, this is the next error when I run the example on https://tensorflow.rstudio.com/tutorials/keras/classification.

```
> model %>% fit(train_images, train_labels, epochs = 5, verbose = 2)
2023-03-09 12:13:49.507435: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
Epoch 1/5
2023-03-09 12:13:49.664416: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:114] Plugin optimizer for device_type GPU is enabled.
2023-03-09 12:13:49.755374: W tensorflow/core/framework/op_kernel.cc:1830] OP_REQUIRES failed at xla_ops.cc:418 : NOT_FOUND: could not find registered platform with id: 0x2b472b5e0
2023-03-09 12:13:49.755404: W tensorflow/core/framework/op_kernel.cc:1830] OP_REQUIRES failed at xla_ops.cc:418 : NOT_FOUND: could not find registered platform with id: 0x2b472b5e0
2023-03-09 12:13:49.766024: W tensorflow/core/framework/op_kernel.cc:1830] OP_REQUIRES failed at xla_ops.cc:418 : NOT_FOUND: could not find registered platform with id: 0x2b472b5e0
2023-03-09 12:13:49.766046: W tensorflow/core/framework/op_kernel.cc:1830] OP_REQUIRES failed at xla_ops.cc:418 : NOT_FOUND: could not find registered platform with id: 0x2b472b5e0
Error: tensorflow.python.framework.errors_impl.NotFoundError: Graph execution error:
<... omitted ...>b/python3.8/site-packages/keras/optimizers/optimizer_experimental/optimizer.py", line 1166, in _internal_apply_gradients
return tf.__internal__.distribute.interim.maybe_merge_call(
File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/optimizers/optimizer_experimental/optimizer.py", line 1216, in _distributed_apply_gradients_fn
distribution.extended.update(
File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/optimizers/optimizer_experimental/optimizer.py", line 1211, in apply_grad_to_update_var
return self._update_step_xla(grad, var, id(self._var_key(var)))
Node: 'StatefulPartitionedCall_2'
could not find registered platform with id: 0x2b472b5e0
[[{{node StatefulPartitionedCall_2}}]] [Op:__inference_train_function_1111]
See `reticulate::py_last_error()` for details
> reticulate::py_last_error()
Traceback (most recent call last):
  File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/tensorflow/python/eager/execute.py", line 52, in quick_execute
    tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,
tensorflow.python.framework.errors_impl.NotFoundError: Graph execution error:

Detected at node 'StatefulPartitionedCall_2' defined at (most recent call last):
    File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 65, in error_handler
      return fn(*args, **kwargs)
    File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/engine/training.py", line 1650, in fit
      tmp_logs = self.train_function(iterator)
    File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/engine/training.py", line 1249, in train_function
      return step_function(self, iterator)
    File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/engine/training.py", line 1233, in step_function
      outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/engine/training.py", line 1222, in run_step
      outputs = model.train_step(data)
    File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/engine/training.py", line 1027, in train_step
      self.optimizer.minimize(loss, self.trainable_variables, tape=tape)
    File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/optimizers/optimizer_experimental/optimizer.py", line 527, in minimize
      self.apply_gradients(grads_and_vars)
    File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/optimizers/optimizer_experimental/optimizer.py", line 1140, in apply_gradients
      return super().apply_gradients(grads_and_vars, name=name)
    File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/optimizers/optimizer_experimental/optimizer.py", line 634, in apply_gradients
      iteration = self._internal_apply_gradients(grads_and_vars)
    File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/optimizers/optimizer_experimental/optimizer.py", line 1166, in _internal_apply_gradients
      return tf.__internal__.distribute.interim.maybe_merge_call(
    File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/optimizers/optimizer_experimental/optimizer.py", line 1216, in _distributed_apply_gradients_fn
      distribution.extended.update(
    File "/Users/jz/Library/r-miniconda-arm64/envs/r-reticulate/lib/python3.8/site-packages/keras/optimizers/optimizer_experimental/optimizer.py", line 1211, in apply_grad_to_update_var
      return self._update_step_xla(grad, var, id(self._var_key(var)))
Node: 'StatefulPartitionedCall_2'
could not find registered platform with id: 0x2b472b5e0
	 [[{{node StatefulPartitionedCall_2}}]] [Op:__inference_train_function_1111]
```

Any hints? Thank you!

**Comment by zjl19902003** [comment]
> 

Thank you!

Not sure if my next question is off-topic, but will similar issues happen to an AMD CPU? Thanks!

**Comment by t-kalinowski** [comment]
The `tensorflow-macos` package is not relevant if you're on an AMD CPU. In that case you're likely running Windows or Linux, and then the standard `tensorflow` package will work just fine. Note, Tensorflow AMD **GPU**s are not supported, if you want to use a GPU in that case, you'll need to use an Nvidia one.

**Comment by LTLA** [comment]
FWIW, the original "spaces in paths" issue should now be fixed in pyenv/pyenv#3249.
